src/commands.cpp

```diff
@@ -1974,60 +1988,178 @@ void CommandGetUA::procresult()
+                client->app->getua_result(API_EINTERNAL);
+                delete [] data;
+                return;
+            }
 
-            // Is the data a multiple of the cipher blocksize, then we're using
-            // a zero IV.
-            if (l % client->key.BLOCKSIZE == 0)
+            if (attributename == "*keyring")
             {
-                if (!PaddedCBC::decrypt(&d, &client->key))
+                // Ed25519 not initialized yet
+                if (!client->signkey)
                 {
-                    delete[] data;
-                    client->app->getua_result(API_EINTERNAL);
-                    return;
+                    if (tlvRecords->find(EdDSA::TLV_KEY))   // use received value
+                    {
+                        client->signkey = new EdDSA((unsigned char *) tlvRecords->get(EdDSA::TLV_KEY).data());
+                    }
+                    else    // create e new keypair
+                    {
+                        client->signkey = new EdDSA();
+                        missingKeypair = true;
+                    }
+                }
+
+                // x25519 not initialized yet
+                if (!client->chatkey)
+                {
+                    if (tlvRecords->find(ECDH::TLV_KEY))   // use received value
+                    {
+                        client->chatkey = new ECDH((unsigned char *) tlvRecords->get(ECDH::TLV_KEY).data());
+                    }
+                    else    // create e new keypair
+                    {
+                        client->chatkey = new ECDH();
+                        missingKeypair = true;
+                    }
+                }
+
+                // any of the keys is missing in MEGA --> update the *keyring
+                if (missingKeypair)
+                {
+                    tlvRecords->set(EdDSA::TLV_KEY, string((char *) client->signkey->keySeed));
+                    tlvRecords->set(ECDH::TLV_KEY, string((char *) client->chatkey->privKey));
+
+                    tlvContainer = tlvRecords->TLVrecordsToContainer();
+                    client->putua(attributename.c_str(), (byte *) tlvContainer.data(), tlvContainer.length());
+                    client->putua("+puEd255", (byte *) client->signkey->pubKey, EdDSA::PUBLIC_KEY_LENGTH);
+                    client->putua("+puCu255", (byte *) client->chatkey->pubKey, ECDH::PUBLIC_KEY_LENGTH);
+                }
+                else    // all good, ready to check public keys
+                {
+                    client->getua(client->finduser(client->me), "+puEd255");
+                    client->getua(client->finduser(client->me), "+puCu255");
                 }
             }
+
+            client->app->getua_result(tlvRecords);
+            delete tlvRecords;
+
+            break;
+
+        case '+':   // public
+
+            // legacy public attributes: '+' prefix but not TLV
```

Comment:
public attributes are **never** by default TLV encoded. So none of these is legacy. So, even `+a` should not be legacy, as it's just containing binary avatar content. The other ones are **definitely* not legacy and fully well behaved.

It's **only** private attributes that are TLV encoded, and are in need of automatic TLV unwrapping.